Email Notification Customization

SnorlaxYum opened the pull request:
## Mail Format Choose
Available options: `plain`, `html`, `multipart`. Default set to `plain`, so you don't need to config this in the mail block of the isso server conf if you mean to use plain text:

	[mail]
	...
	format = plain

To use multipart, set the option to `multipart` in the mail block of the isso server conf:

	[mail]
	...
	format = multipart

To use html, set the option to `html` in the mail block of the isso server conf:

	[mail]
	...
	format = html

## Title Customization  

To customize the mail title, set like the following options in the mail block of the isso server conf:  

	[mail]
	...
	title_admin = {replier} commented on {title}
	title_user = {receiver}, {replier} replied your comments on {title}

**Result:**  
![](https://static.snorl.ax/isso-noti/20181230/def/url_mod.png)
![](http://static.snorl.ax/isso-noti/20181230/def/reply_url.png)

## Mail Language

	[mail]
	...
	language = en

https://github.com/posativ/isso/pull/518#issuecomment-450546330

## Mail Customization  

All three formats support customization, it helps to read the dafault configuration for details (`isso/templates/comment.html` for `format = html`, `isso/templates/comment.plain` for `format = plain`).

To use the feature, set the following options like this in the mail block of the isso server conf:   

	[mail]
	...
	template = /the/path/to/the/directory/or/a-single-template

If you set it to the path to a directory, 
for `format = html` it will check if `admin.html` and `user.html` in that directory are available. 
(For `format = plain` it will check `admin.plain` and `user.plain`; 
For `format = multipart`, it will check `admin.html` and `user.html` for html part, `admin.plain` and `user.plain` for plain part) 
The name of the file should be self-explaining. `user.html` and `user.plain` are used for reply notifications to the subcribed users, `admin.html` and `admin.plain` are used for mail notifications to the admin.  
If one of these is missing then in the corresponding case the template will fall back to the default.
The log will tell you what happened when the error occurs. The template will be rendered by Jinja2.

Alternatively, you could just directly edit the default mail template without setting the option `template`.

Default template path for `format = plain` (Different if you set a different mail language using the option `language`):  
When mail language is not set using `language` or set via `language = en`: `isso/templates/comment.plain`
When it's set any other language  than `en` using `language`, take `language = ja` for example:  `isso/templates/comment_ja.plain`

Default template path for `format = html` (Different if you set a different mail language using the option `language`):  
When mail language is not set using `language` or set via `language = en`: `isso/templates/comment.html`
When it's set any other language  than `en` using `language`, take `language = ja` for example:  `isso/templates/comment_ja.html`

Default template path for `format = multipart`:  
html part: the same as Default template path for `format = html` (listed above)
plain part: the same as Default template path for `format = plain` (listed above)

Templating Examples: https://github.com/posativ/isso/pull/518#issuecomment-450542453

__HTML Styling suggestion:__  Using inline css as much as u can, cause webmail service like gmail will block any kind of css on the web other than the inline css. In imap `<style>` tag works in mail. Don't leave out the `<html>` out as this will add to spam score to your mail.  

__Format choose suggestion:__ Don't use `format = html`, the mail will have a higher spam score for having only html part. If you wanna use html, use `format = multipart` instead.  

**Plain Text Customization Suggestion:** remember to tidy your text and remove every unnecessary spaces besides the comment you want (things like `{#...#}` won't be rendered), 'cause they will be rendered into the final mail. 

**Fallback Template:** https://github.com/posativ/isso/pull/518#issuecomment-450537644

SnorlaxYum commented:
Result In Mail:
Reply Notification:  
![Reply Notification](https://static.snorl.ax/isso-noti/reply-noti.png)
The comment published without moderation and user's url:  
![](https://static.snorl.ax/isso-noti/direct-nourl.png)
The comment directly published with user's url:  
![](https://static.snorl.ax/isso-noti/direct-url.png)
The comment waiting for approval without user's url:  
![](https://static.snorl.ax/isso-noti/mod-nourl.png)
The comment waiting for approval with user's url:  
![](https://static.snorl.ax/isso-noti/mod-url.png)

SnorlaxYum commented:
If u don't fill in any of the 5 options, they will fall back to the default format isso uses.

SnorlaxYum commented:
- [ ] [Easy] Add a author name variable without email so the `Sim<email>'s URL` could be `Sim's URL`
- [ ] [Easy] Add The Config To Documentation
- [ ] [Trying To Find A Way] Add them to the config file so `WARNING: no such option` won't pop.

SnorlaxYum commented:
__Fallback Default when u don't set any of the 6 options (The Same As the default before. The website URL of the replier won't appear in reply notification even if he sets it):__

Reply Notification when replier has an URL:  
![Reply Notification](https://static.snorl.ax/isso-noti/20181230/default/url_rep.png)
Reply Notification when replier doesn't submit an URL:  
![Reply Notification](https://static.snorl.ax/isso-noti/20181230/default/nourl_rep.png)
The comment published without moderation and user's url:  
![](https://static.snorl.ax/isso-noti/20181230/default/nourl_direct.png)
The comment directly published with user's url:  
![](https://static.snorl.ax/isso-noti/20181230/default/url_direct.png)
The comment waiting for approval without user's url:  
![](https://static.snorl.ax/isso-noti/20181230/default/nourl_mod.png)
The comment waiting for approval with user's url:  
![](https://static.snorl.ax/isso-noti/20181230/default/url_mod.png)

SnorlaxYum commented:
- [ ] [Easy]Title Customization

SnorlaxYum commented:
To customize the mail title, set like the following options in the smtp block of the isso server conf:
	[smtp]
	...
	mail_title_admin = {replier} commented on {title}
	mail_title_user = {receiver}, {replier} replied your comments on {title}
Test result:  
![](https://static.snorl.ax/isso-noti/20181230/title_admin.png)
![](https://static.snorl.ax/isso-noti/20181230/title_user.png)

matolivier commented:
Hi, why not adding the user website url as default behaviour in Isso and adding international translation for the text in current code?

SnorlaxYum commented:
> Hi, why not adding the user website url as default behaviour in Isso and adding international translation for the text in current code?

I didn't change the default setting 'cause this may conflict with some users who are already satisfied with the default setting.
I'll consider the translation after doing this:
- [ ] [Easy] Add ability to add parent comment URL to the notification url.

SnorlaxYum commented:
Now it's able to add parent link to reply notification with `{parent_link}` tag.  

	user_format_url = {author} replied to your comment:
	
	        {comment}
	
	        Your comment: {parent_link}
	        {only_author}'s URL: {website}
	        Link to the reply: {link}
	        ---
	        Unsubscribe from this conversation: {unsubscribe}
	user_format_nourl = {author} replied to your comment:
	
	        {comment}
	
	        Your comment: {parent_link}
	        Link to the reply: {link}
	        ---
	        Unsubscribe from this conversation: {unsubscribe}

The Test Result in Reply Notification Mail:  
![](https://static.snorl.ax/isso-noti/20181230/pa_url.png)
![](https://static.snorl.ax/isso-noti/20181230/pa_nourl.png)

SnorlaxYum commented:
To do:  
- [ ] Translation Ability

@matolivier What's the language u need? Or u do the translation of ur language for the fallback mail format?

matolivier commented:
@SnorlaxYum, I apologize for my raw answer earlier which I should have expanded. Indeed the fact to change the text thanks to your work is an advantage, however the idea of translation is having the emails text ready to use for anyone instead of having to translate manually in conf.

I would need french translation (I can provide the text). Maybe I could take a look on failback mecanism (the one that is curre tly used).

SnorlaxYum commented:
> @SnorlaxYum, I apologize for my raw answer earlier which I should have expanded. Indeed the fact to change the text thanks to your work is an advantage, however the idea of translation is having the emails text ready to use for anyone instead of having to translate manually in conf. ;)
> 
> I would need french translation (I can provide the text). Maybe I could take a look on failback mecanism (the one that is curre tly used).

Okay, translate the value of these variables:  

    admin_format_urluser_moderate = 
        {author} wrote:
        
        {comment}
        
        User's URL: {website}
        IP address: {ip}
        Link to comment: {com_link}
        
        ---
        Delete comment: {del_link}
        Activate comment: {act_link}
    
    admin_format_nourluser_moderate = 
        {author} wrote:
        
        {comment}
        
        IP address: {ip}
        Link to comment: {com_link}
        
        ---
        Delete comment: {del_link}
        Activate comment: {act_link}
    
    admin_format_urluser_direct = 
        {author} wrote:
        
        {comment}
        
        User's URL: {website}
        IP address: {ip}
        Link to comment: {com_link}
        
        ---
        Delete comment: {del_link}
    
    admin_format_nourluser_direct = 
        {author} wrote:
        
        {comment}
        
        IP address: {ip}
        Link to comment: {com_link}
        
        ---
        Delete comment: {del_link}
    
    user_format_url = 
        {author} wrote:
        
        {comment}
        
        Link to comment: {link}
        
        ---
        Unsubcribe from this conversation: {unsubscribe}
        
    user_format_nourl = 
        {author} wrote:
        
        {comment}
        
        Link to comment: {link}
        
        ---
        Unsubcribe from this conversation: {unsubscribe}

Also, the translation for the word "Anonymous" in French is needed.  
The translated values will be used for `anonymous_fr`, `admin_format_urluser_moderate_fr`, `admin_format_nourluser_moderate_fr`, `admin_format_urluser_direct_fr`, `admin_format_nourluser_direct_fr`, `user_format_url_fr`, `user_format_nourl_fr`.  
Thanks for ur help

SnorlaxYum commented:
> @SnorlaxYum, I apologize for my raw answer earlier which I should have expanded. Indeed the fact to change the text thanks to your work is an advantage, however the idea of translation is having the emails text ready to use for anyone instead of having to translate manually in conf. ;)
> 
> I would need french translation (I can provide the text). Maybe I could take a look on failback mecanism (the one that is curre tly used).

BTW, you don't need to translate the tag wrapped in `{` and `}`, leave the way as it is.
No need for the variable names like `user_format_nourl` as well. They will be all trailed by `_fr` then.  
Just like this: `user_format_nourl` -> `user_format_nourl_fr`

SnorlaxYum commented:
Now it's possible to translate.  
Just set `mail_language` in `[smtp]` section could do the trick.

	mail_language = zh

![](https://static.snorl.ax/isso-noti/20181230/zh_ad.png)
![](https://static.snorl.ax/isso-noti/20181230/zh_re.png)
Tittle's no big deal, it's my setting overriding it.

SnorlaxYum commented:
BTW, now `mail_language` is only supporting `zh` and `en`, maybe u can help me with the translation of other languages?

SnorlaxYum commented:
The checklist of the languages isso script has already got:
- [ ] bg
- [ ] cs
- [ ] da
- [ ] de
- [ ] el
- [ ] en
- [ ] es
- [ ] eo
- [ ] fa
- [ ] fi
- [ ] fr
- [ ] hr
- [ ] hu
- [ ] it
- [ ] sv
- [ ] nl
- [ ] vi
- [ ] ru
- [ ] zh
- [ ] zh_CN
- [ ] zh_TW

The one whose checkbox is already ticked is now supportted by the `mail_language` option.

jelmer commented on docs/docs/configuration/server.rst:
```diff
@@ -229,6 +229,84 @@ timeout
+	is the title of the thread. Available tags: `{title}`, `{receiver}`,
+	`{replier}`
+
+admin_format_urluser_moderate
```
Can we perhaps use a real templating system for the e-mails, e.g. jinja2 ? That way, we don't end up with a plethora of configuration variables, one for each combination of tags that is available.

SnorlaxYum replied:
Okay, I'll try looking into that. I also deem that plain text is not really flexible.

matolivier commented:
Nice work! It is much cleaner this way.
I will commit a french template as soon as I end my vacations this week.

SnorlaxYum commented:
After some silly mistakes, now `{{thread_link}}` and `{{thread_title}}` is ready for use in template:  
![](https://static.snorl.ax/isso-noti/20181231/thread_link_title.png)

SnorlaxYum commented:
> Nice work! It is much cleaner this way.
> I will commit a french template as soon as I end my vacations this week.

Thank u, the default template is located in ` isso/templates/comment.html`, after having done that, it's fine to make a pull request to my notification branch add the French-translated template named  `isso/templates/comment_fr.html`.

jelmer commented:
Thanks! Can you please add separate templates for the different cases, possibly having each include a common base file? Having to edit a giant file that just references numbers is not very user friendly.

Could we stick to plain text e-mail, or if we have to add HTML e-mail, also ship a plain text version?

SnorlaxYum commented:
Okay I'll add an option to opt between html and plain text version.
You mean a base template and several for  each case based on the base html? I guess it's possible with the current mail_template option, it could be an option for base template directory, I'll separate the template into each one for each case and change the config way.

On December 31, 2018 4:02:05 PM GMT+08:00, "Jelmer Vernooĳ" <notifications@github.com> wrote:
>Thanks! Can you please add separate templates for the different cases,
>possibly having each include a common base file? Having to edit a giant
>file that just references numbers is not very user friendly.
>
>Could we stick to plain text e-mail, or if we have to add HTML e-mail,
>also ship a plain text version?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/posativ/isso/pull/518#issuecomment-450617961

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

SnorlaxYum commented:
- [ ] (templating) separate template for each case  
- [ ] (option) add an option to toggle between plain text and html  
- [ ] (templating) add customization feature for plain text and add docs for it.

SnorlaxYum commented:
Seperate template is available now, I'll write the doc about this later.

matolivier commented:
Hi @SnorlaxYum, as discussed I pushed you a PR for french translations.

SnorlaxYum commented:
Nice @matoliver thanks😆

matolivier commented:
Hello, don't forget about tiking "fr" checkbox.

SnorlaxYum commented:
> Hello, don't forget about tiking "fr" checkbox.

Thanks for notifying.😂
Just ticked

jelmer commented:
Sorry for the slow reply. I think this is much better, but can you please remove most of the configuration options here? It's adding significant complexity to the code and without a clear benefit (i.e. the single digit percentage of users that want to modify the templates can just edit them).

SnorlaxYum commented:
> Sorry for the slow reply. I think this is much better, but can you please remove most of the configuration options here? It's adding significant complexity to the code and without a clear benefit (i.e. the single digit percentage of users that want to modify the templates can just edit them).

This is what I just added to the top of this request:  

> Alternatively, you could just directly edit the default mail template without setting the option `mail_template`.
> 
> Default template path for `mail_format = plain` (Different if you set a different mail language using the option `mail_language`):  
> When mail language is not set using `mail_language` or set via `mail_language = en`: `isso/templates/comment.plain`
> When it's set any other language  than `en` using `mail_language`, take `mail_language = ja` for example:  `isso/templates/comment_ja.plain`
> 
> Default template path for `mail_format = html` (Different if you set a different mail language using the option `mail_language`):  
> When mail language is not set using `mail_language` or set via `mail_language = en`: `isso/templates/comment.html`
> When it's set any other language  than `en` using `mail_language`, take `mail_language = ja` for example:  `isso/templates/comment_ja.html`

Altogether, I think I've added these options:  

* `mail_format`
* `mail_title_admin`
* `mail_title_user`
* `mail_template`
* `mail_language`
* `anonymous` and `anonymous_xx` for various `mail_language` (which I didn't mention in the doc for the sake of complexity)

There's no need in deleting the first 4 options.  
`mail_template` seems a little optional (However it adds convenience if you update via `git pull` and forget to put the edited file in `.gitignore`, the custom path u set won't be affected if the path is out of the source directory).  
`anonymous` is okay for me, adding to customizability even when another language is used.

Actually the six options can be optional. They all have fallback values. Maybe add another section for `Mail Customization` and include the six options?

SnorlaxYum commented:
> The issue I have with it is that it adds a lot of extra code paths that we'll need to keep working in the future. It's very easy for one of these code paths to regress since isso doesn't have a lot of unit tests - as we saw with 0.12.0.
> 
> If you insist on keeping these variables, then can you please:
> 
>     * add unit tests for each one of them
> 
>     * move them to a different section than "smtp"; perhaps "mail" or "notifications" ?

Sorry for my late reply. Just got back to this pull request and add one format `multipart`, 'cause a mail will have higher spam score in spamassassin for having only html part. `multipart` is the perfect solution for that.

I have moved the options to a new section `[mail]`, and wondering the unit tests you say. I guess you mean something like this in `def __init__` of `class SMTP`:  

        # test SMTP connectivity
        try:
            with SMTPConnection(self.conf):
                logger.info("connected to SMTP server")
        except (socket.error, smtplib.SMTPException):
            logger.exception("unable to connect to SMTP server")

In `def format` of `class SMTP`, I actually write something for debuging like this:  

        if self.mail_lang != "en":
            com_ori = os.path.join(temp_path, "comment_%s.%s" % (self.mail_lang, part))
            try:
                jinjaenv.get_template(com_ori)
            except jinja2_exceptions.TemplateSyntaxError as err:
                logger.warn("[mail] Wrong format. %s"%err)
                logger.warn("[mail] Default template fell back to the one for en.")
            except jinja2_exceptions.TemplateNotFound:
                logger.warn("[mail] No default template for such language: %s. Default template fell back to the one for en." % self.mail_lang)
            except Exception as err:
                logger.warn("[mail] Some error about jinja2. %s"  % type(err))
                for er in err.args:
                    logger.warn(      "%s" % er)
                logger.warn("[mail] Default template fell back to the one for en.")
            else:
                com_ori_admin = com_ori_user = os.path.basename(com_ori)

Or you mean something like `test_*.py` in `isso/tests/`?

jelmer commented:
> Or you mean something like `test_*.py` in `isso/tests/`?

Yep, those are the unit tests.

matolivier commented:
Hi there, any news? With the time going the PR may have more and more conflicts.

matolivier commented:
Hi @SnorlaxYum, could you take a look at jelmer reviews please, so the PR could progress?

SnorlaxYum commented on share/isso.conf:
```diff
@@ -133,6 +133,58 @@ from =
+mail_language = en
+
+# specify mail format to use, available options: plain, html. Default: plain
+mail_format = plain
```
It's outdated, now multipart is supported.

SnorlaxYum replied:
Okay, it's my fault, I forgot to add the multipart to the comment

SnorlaxYum commented on docs/docs/configuration/server.rst:
```diff
@@ -229,6 +229,37 @@ timeout
+	code, default to `en`.
+	
+mail_format
+	specify mail format to use, available options: plain, html. Default: plain
```
As I said, multipart is supported.

SnorlaxYum commented on isso/ext/notifications.py:
```diff
@@ -79,7 +82,16 @@ def __init__(self, isso):
+        self.mail_format = self.isso.conf.get("smtp", "mail_format")
+        
+        try:
+            self.no_name = self.isso.conf.get("smtp", "anonymous_%s" % self.mail_lang)
```
So you mean all the anonymous_lang option should be eliminated then?

SnorlaxYum commented on docs/docs/configuration/server.rst:
```diff
@@ -229,6 +229,37 @@ timeout
+	is the title of the thread. Available variables: `{title}`, `{receiver}`,
+	`{replier}`
+	
+mail_template
```
It's possible, I add this just for customization.

SnorlaxYum commented:
> Hi @SnorlaxYum, could you take a look at jelmer reviews please, so the PR could progress?

Sorry for late reply, just did that.

@jelmer, sorry for the late reply, maybe take a look at the latest commit? I just eliminated the anonymous options. The reason I have `mail_template` remained here is for customization. Do we have to eliminate that?

jelmer commented:
Thanks; I don't see that the last commit addresses the comments - the unit tests still just test the ini parser, not the logic that you're adding.

SnorlaxYum commented:
> Thanks; I don't see that the last commit addresses the comments - the unit tests still just test the ini parser, not the logic that you're adding.

Sorry, but I'm indeed somewhat confused about what you said. 
What I added was mainly some format things using python string format and jinja2, there is no problem about the format function itself, the mail sending function is just working like before. 
I am really wondering what the unit test is used for.

jelmer commented:
I'm not sure how to phrase this differently - the unit tests should cover the logic that you've changed *in notifications.py* - and my reviews of this pull request have all been about that. Even if all the code paths you've added have been tested manually now, changes in the future can break them and people will only notice if they break the code paths that are normally used (i.e. when nothing special is used - i.e. not "multipart").

SnorlaxYum commented:
> I'm not sure how to phrase this differently - the unit tests should cover the logic that you've changed _in notifications.py_ - and my reviews of this pull request so far have all emphasized that. You're making notifications.py significantly more complex. Even if all the code paths you've added have been tested manually now, changes in the future can break them and people will only notice if they break the code paths that are normally used (i.e. when nothing special is used - i.e. not "multipart").

Helo, these are the logic of the things I added:  
The modules I edited in `SMTP` class:  
1. `__init__`:   
add self.mail_lang to get the language setting.  
add self.mail_format to get the setting of the mail type(plain/html/multipart)  
add self.no_name_list to set "Anonymous"(self.no_name) according to the language.  
select from self.no_name_list according to self.mail_lang to set self.no_name, fallback to "Anonymous" when the corresponding string is not found.  

2. `format`:  
Mainly change the way the mail is formated.  
Add an argument `part` to determine which default template to use: Use the html template when the argument `part` received is "html", use the plain template when the argument `part` received is "plain". self.mail_lang is used here to determine which template to use.   

3. `notify_new`:  
add a case when self.mail_format is set to multipart where the message should contain both html and plain part (format twice, once in plain, once in html).  

4. `notify_users`:  
replace "Anonymous" with self.no_name.  
add a case when self.mail_format is set to multipart where the message should contain both html and plain part (format twice, once in plain, once in html).  

5. `sendmail`, `_retry`:  
add a case when self.mail_format is set to multipart where the message should contain both html and plain part (format twice, once in plain, once in html).  

6. `_sendmail`:  
replace the original "plain" with self.mail_format to enable the possibility to choose from "html" and "plain"
add a case when self.mail_format is set to "multipart" where the message should contain both html and plain part (format twice, once in plain, once in html).

All logger things added make it easier for users to debug.  
They are actually are things mainly about the way the mail is formated and don't change the way the mail is sent.  
What is a unit test aimed to do here? To test already tested formatting function or mail sending function?

jelmer commented:
On Sun, Apr 07, 2019 at 03:29:11AM -0700, Sim wrote:
> They are actually are things mainly about the way the mail is formated and don't change the way the mail is sent.  
> What is a unit test aimed to do here? To test already tested formatting function or mail sending function?
The unit tests are meant to verify that all of the code paths that
you're adding are working correctly.

Does it do the right thing when mail_lang!="en"? Does it do the right
thing if template doesn't render? Does it do the right thing if
the path specified is a directory, and if it's a file?

For regular users, only one of those paths is ever going to be
exercised. Unit tests should help make sure that the other code paths
don't break as other bits of isso are being changed.

Jelmer

-- 
Jelmer Vernooĳ <jelmer@jelmer.uk>
PGP Key: https://www.jelmer.uk/D729A457.asc

SnorlaxYum commented:
> On Sun, Apr 07, 2019 at 03:29:11AM -0700, Sim wrote: They are actually are things mainly about the way the mail is formated and don't change the way the mail is sent. What is a unit test aimed to do here? To test already tested formatting function or mail sending function?
> The unit tests are meant to verify that all of the code paths that you're adding are working correctly. Does it do the right thing when mail_lang!="en"? Does it do the right thing if template doesn't render? Does it do the right thing if the path specified is a directory, and if it's a file? For regular users, only one of those paths is ever going to be exercised. Unit tests should help make sure that the other code paths don't break as other bits of isso are being changed. Jelmer
> […](#)
> -- Jelmer Vernooĳ <jelmer@jelmer.uk> PGP Key: https://www.jelmer.uk/D729A457.asc

Okay, I guess you mean testing all the possible cases.  
I'll write unit tests to test if the formatted content matches the expected result in all the cases I could think of.

SnorlaxYum commented:
@jelmer any feedback?

jelmer commented:
Thanks! These aren't really unit tests, but more integration tests - but that's fine too (though less preferred), so long as they can be run without user interaction. I've left a bunch of inline comments on some things I spotted quickly, I'll try and take some more time to respond in more detail later this week.

Please follow PEP8 and make sure that the tests pass; you can check the errors by running "make flakes test".  We used to have travis enabled for this repository; I'm not sure what happened with that..

SnorlaxYum commented on isso/ext/notifications.py:
```diff
@@ -104,52 +143,125 @@ def __iter__(self):
+                logger.warn("[mail] Some error about jinja2. %s"  % type(err))
+                for er in err.args:
+                    logger.warn("      %s" % er)
+                logger.warn("[mail] Default template fell back to the one for en.")
```
No need to verify 'cause it won't change the default temp which is en. But it's okay if you mean adding a test for it.

SnorlaxYum commented on isso/tests/test_mail.py:
```diff
@@ -0,0 +1,268 @@
+
+from isso.ext.notifications import SMTP
+
+smtp_host = input("Enter the SMTP host (Leave it blank to use localhost): ")
```
I'll eliminate these.  
But the error can't be eliminated when no SMTP service is running on port 587:  
ConnectionRefusedError: [Errno 111] Connection refused  

Considering the error doesn't affect the result of the unit test, guess that's not a big problem.

SnorlaxYum commented:
[ ] Translation system for Anonymous
[ ] Test for the language not available
[ ] Eliminate the trailing slash

SnorlaxYum commented:
@jelmer Just added a few unit tests, used translate package for Anonymous instead, thus adding a dependency `translate` to isso. Any feedback?

SnorlaxYum commented:
@jelmer Hello, I've updated in case that Travis doesn't get the value. Now it seems there's problem with the Travis CI itself: https://travis-ci.org/posativ/isso/jobs/520726111  

The error message:  

        rm: cannot remove '/dev/shm': Device or resource busy
        The command "sudo rm -rf /dev/shm && sudo ln -s /run/shm /dev/shm" failed and exited with 1 during .

Seems the problem lies in the virtual machine itself?

Regardless of that, I'm gonna update the branch to make sure it's up to date, the changes that happened to the mail template I've found:  

- [ ] Include user e-mails in admin notifications only https://github.com/posativ/isso/pull/535

SnorlaxYum commented:
@jelmer thanks for your help, now all checks have passed. I just enriched the doc for it to ensure:  

* This pull request is included in the docs for reference
* How to contribute to the translation for the templates is included in the docs.  

Any feedback?

SnorlaxYum commented:
Sorry for the late update, since I discovered this is needed today, so I just added some codes that convert the original markdown code to html code in html part of the mail, here's the preview of the conversion:  
![Mail markdown to html](https://static.snorl.ax/isso-noti/mail-markup-html-preview.png)

The supported markdown code is based on how you set [Markup](https://posativ.org/isso/docs/configuration/server/#markup) section in the isso configuration file.  

The relevant Markup section setting about the example above:  
```
[markup]
options = strikethrough, fenced-code, footnotes, superscript, autolink
allowed-elements = img
allowed-attributes = src
```

SnorlaxYum commented:
Any feedback or suggestions? @jelmer @matolivier @silberzwiebel @molivier-fcms

matolivier commented:
Sorry @SnorlaxYum, I barely have the time to do my own stuffs, but I'll try to find a time for testing the changes.

jelmer commented on docs/docs/configuration/server.rst:
```diff
@@ -229,6 +229,75 @@ timeout
+Mail
+----
+
+Confiurgure the format the mail notification uses in SMTP section.
```
Configure

jelmer commented on isso/tests/test_mail.py:
```diff
@@ -0,0 +1,342 @@
+from isso import Isso, core, config, local, dist
+from isso.utils import http
+from fixtures import curl, loads, FakeIP, JSONClient
+http.curl = curl
```
Where does http.curl get reset?

jelmer commented on isso/tests/test_mail.py:
```diff
@@ -0,0 +1,342 @@
+
+
+class TestMail(unittest.TestCase):
+    def setUp(self):
```
Missing empty line per PEP8

jelmer commented on isso/tests/test_mail.py:
```diff
@@ -0,0 +1,342 @@
+
+class TestMail(unittest.TestCase):
+    def setUp(self):
+        fd, self.path = tempfile.mkstemp()
```
Please clean up the file at the end of the test.

jelmer commented on isso/tests/test_mail.py:
```diff
@@ -0,0 +1,342 @@
+
+        self.smtp = SMTP(self.app)
+
+# Test if anonymous works
```
This seems like it should be a docstring on the function?

jelmer commented on isso/tests/test_mail.py:
```diff
@@ -0,0 +1,342 @@
+                             del_link=self.public_endpoint + "/id/%i/delete/" % rv["id"] + self.smtp.key,
+                             com_link=local("origin") + thread_test["uri"] + "#isso-%i" % rv["id"]))
+
+# Test what will happen when these happens:
```
This should be a docstring.

jelmer commented on isso/tests/test_mail.py:
```diff
@@ -0,0 +1,342 @@
+# 2. The default template could not be found.
+
+    def testAnonymousStringNotAvailable(self):
+        self.conf.set("mail", "languagFe", "aa")
```
Typo?

jelmer commented on share/isso.conf:
```diff
@@ -134,6 +134,61 @@ from =
+# specify mail format to use, available options: plain, html, multipart. Default: plain
+format = plain
+
+# specify the title format of the notification email sent to the admin.
```
Please document what the possible substitution variables are.

jelmer commented on isso/tests/test_mail.py:
```diff
@@ -0,0 +1,342 @@
+# -*- encoding: utf-8 -*-
```
There don't appear to be any tests for format=html or format=multipart?

SnorlaxYum commented:
@jelmer 
Comment 1: That logic: skip the SMTP test when the module is run by `test_mail.py`
Comment 2: Yes, you are right. And codes right behind that print the err:  
```python
                    for er in err.args:
                        logger.warn("      %s" % er)
```
Comment 3~4: Okay.
Comment 5: I think that's necessary for isso to run, as you can see in `test_comments.py`.
Comment 6: Okay.  
Comment 7: Okay, I will add this after the `setUp` function:  

```
    def tearDown(self):
        os.unlink(self.path)
```

Comment 8~9: Okay, will change them to `"""The description"""` and move them into the function. 
Comment 10: Yes......😅
Comment 11: I thought they are self-explainary😅Okay, I will add some explanation about it.
Comment 12: Yes 'cause I thought jinja2 should render `html` version fine if all the tests on `plain` passes. But I'll add some about `html` as well, `multipart` is a combination of `plain` and `html` so it will definitely work fine if both `plain` and `html` are working well.  

To-do list:  

- [ ] "Configure" to "configure" in the docs
- [ ] In `test_mail.py` add empty line behind the class declarification
- [ ] Add the `tearDown` to the `test_mail.py`
- [ ] Change the comments in `test_mail.py` to docstrings
- [ ] Fix the typo in `test_mail.py`
- [ ] Add the explanation of the var to the description of `title_user`, `title_admin`
- [ ] Add tests for `html` in `test_mail.py`

SnorlaxYum commented:
@jelmer I've just updated it with the promised changes. Any feedbacks?

SnorlaxYum commented:
Fixed some errors, now it should work fine when uwsgi is implemented.

jelmer commented on docs/docs/configuration/server.rst:
```diff
@@ -229,6 +229,86 @@ timeout
+Mail
+----
+
+confiurgure the format the mail notification uses in SMTP section.
```
configure

jelmer commented on docs/docs/configuration/server.rst:
```diff
@@ -229,6 +229,86 @@ timeout
+	specify the title format of the notification email sent to the admin.
+  Default: `{title}`
+  Available variables:
+  `{title}`: The name of the isso thread.
```
s/name/title/ ?

jelmer commented on docs/docs/configuration/server.rst:
```diff
@@ -229,6 +229,86 @@ timeout
+  Default: `{title}`
+  Available variables:
+  `{title}`: The name of the isso thread.
+  `{replier}`: The author name of the comment. If the author didn't fill in the Name
```
s/author name of the comment/name of the author of the comment/

jelmer commented on docs/docs/configuration/server.rst:
```diff
@@ -229,6 +229,86 @@ timeout
+  Available variables:
+  `{title}`: The name of the isso thread.
+  `{replier}`: The author name of the comment. If the author didn't fill in the Name
+  blank, it will fallback to "Anonymous" or something else according to the language
```
double negative and unnecessary capitalization - s/didn't fill in the Name blank/didn't fill in a name/

jelmer commented on docs/docs/configuration/server.rst:
```diff
@@ -229,6 +229,86 @@ timeout
+  else according to the language you set to in the `language` option above.
+
+template
+	specify the path to your customized html template of comment yourself, in
```
s/yourself//

jelmer commented on docs/docs/configuration/server.rst:
```diff
@@ -229,6 +229,86 @@ timeout
+  blank, it will fallback to "Anonymous" or something else according to the language
+  you set to in the `language` option above.
+  `{receiver}`: The author whose comment is replied to in this situation.  If the
+  author didn't fill in the `Name` blank, it will fallback to "Anonymous" or something
```
double negative again.

jelmer commented on isso/templates/comment.txt:
```diff
@@ -0,0 +1,55 @@
+========================================================================
+Variables usable under all situations:
+========================================================================
+{{thread_link}} : self-explainary
```
*self-explanatory

SnorlaxYum commented:
Sorry fot the typos:sweat_smile:, just fixed them and also made `isso/templates/comment.txt` more specified.

jelmer commented on docs/docs/configuration/server.rst:
```diff
@@ -229,6 +229,86 @@ timeout
+
+language
+	specify the language of the mail format by its `ISO 639-1<https://en.wikipedia.org/wiki/ISO_639-1>`_ (two letter)
+	code, default to `en`. The "Anonymous" will be translated in the language you
```
s/The "Anonymous"/The term "Anonymous"

jelmer replied:
s/translated in/translated to/
or
s/translated//

jelmer commented on docs/docs/configuration/server.rst:
```diff
@@ -229,6 +229,86 @@ timeout
+language
+	specify the language of the mail format by its `ISO 639-1<https://en.wikipedia.org/wiki/ISO_639-1>`_ (two letter)
+	code, default to `en`. The "Anonymous" will be translated in the language you
+	set. If you use the default template, the template will also be translated into
```
See above.

jelmer commented on docs/docs/configuration/server.rst:
```diff
@@ -229,6 +229,86 @@ timeout
+	code, default to `en`. The "Anonymous" will be translated in the language you
+	set. If you use the default template, the template will also be translated into
+	the language (Note that the default templates supported to be translated using
+	the option now includes: `en`, `de`, `fr`, `ja`, `zh`, `zh-CN`, `zh-TW`. You are welcome to
```
"Note that the default templates support the following languages:"

But please don't list the languages here, since that means we'll have to keep this list synced with the actual supported languages.

jelmer commented on docs/docs/configuration/server.rst:
```diff
@@ -229,6 +229,86 @@ timeout
+
+title_admin
+	specify the title format of the notification email sent to the admin.
+  Default: `{title}`
```
Broken indentation.

jelmer commented on docs/docs/configuration/server.rst:
```diff
@@ -229,6 +229,86 @@ timeout
+title_user
+	specify the title format of the notification email sent to the subscribed
+	commenter.
+  Default: `Re: New comment posted on {title}`
```
Broken/inconsistent indeptation.

jelmer commented on docs/docs/configuration/server.rst:
```diff
@@ -229,6 +229,86 @@ timeout
+	Default: plain
+
+title_admin
+	specify the title format of the notification email sent to the admin.
```
Do you mean title or subject here?

jelmer commented on docs/docs/configuration/server.rst:
```diff
@@ -229,6 +229,86 @@ timeout
+  else according to the language you set to in the `language` option above.
+
+template
+	specify the path to your customized template of comment, in jinja2 format.
```
Do all these various ways of specifying the format have tests?

jelmer commented on isso/ext/notifications.py:
```diff
@@ -3,16 +3,20 @@
 from email.mime.text import MIMEText
+from jinja2 import Environment, FileSystemLoader
+import jinja2.exceptions as jinja2_exceptions
+from translate import Translator
```
Can we please do this without an extra dependency? All that's needed is a dictionary.

jelmer replied:
As far as I can tell this will actually make HTTP calls.

jelmer commented:
Sorry, I'm running out of patience with this pull request a bit, there are still a lot of things that need to be fixed. :-/ Can I perhaps suggest adding this in incremental steps so we can at least land some of your changes? I.e. make the e-mail subject configurable in the first pull request, then add support for a template for the email body, then add support for multipart, etc?

SnorlaxYum commented:
Comment 1-6: Okay?
Comment 7: Yes, need change on the variable name?
Comment 8: Yes...`/def testCustomTemplate`, okay, a case when dir is not avaible is missing.  
Comment 9: About the `translate`, dict was my choice back then, you suggested using a tranlation system instead.  
![](https://static.snorl.ax/isso-noti/sth/2019-05-29-094409_1920x1080_scrot.png)

> Sorry, I'm running out of patience with this pull request a bit, there are still a lot of things that need to be fixed. :-/ Can I perhaps suggest adding this in incremental steps so we can at least land some of your changes? I.e. make the e-mail subject configurable in the first pull request, then add support for a template for the email body, then add support for multipart, etc?

Sorry for my inefficiency.  
You mean seperate PRs? Okay.

To-do list:  
[ ] Seperate things into several PRs.  

PR separation:  
[ ] PR 1: Email subject customization.
[ ] PR 2: Add support for html body in email.
[ ] PR 3: Add support for multipart in email. (Will be created once PR 2 is merged)

SnorlaxYum commented:
Okay, the PR is now closed due to a lot of messes.  
I'll create separate PRs in the future about this PR.  
If you wanna try the feature in this PR to help me with my future PRs, you can try the [`notification` branch](https://github.com/SnorlaxYum/isso/tree/notification).

jelmer commented:
With regard to the translation system, I mentioned "the translation system" - not a random translation system that will contact Microsoft to translate a single word.

Isso has an existing translation system in isso/js/app/i18n/; it's intended for JavaScript, but probably makes sense to also make it work for Python.

SnorlaxYum commented:
Okay, I Will try to integrate with that. Thank you for your patience.  

To-do:  

-[ ] Re-edit variables in templates, since I spotted some errors about my understanding for repliee.
-[ ] Integrate with the existing translation system in isso.  

This branch will continue as an upstream for the newly added mail module though closed for PR.

vegeta123254 commented:
Sorry @SnorlaxYum , Is this feature available?

SnorlaxYum commented:
The PR hasn't been merged into the main branch.

So it's definitely not available in the official versions of isso.

And I don't think it's ever gonna be merged.

That's why I closed the PR.

I'm using this modified branch on my site though.
